File: app/src/main/java/com/jsb/calculator/activity/CustomizeKeyboardActivity.java
package com.jsb.calculator.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.TextView;
import android.widget.Toast;

import com.jsb.calculator.Classes.Fonts;
import com.jsb.calculator.R;
import com.jsb.calculator.databinding.ActivityCustomizeKeyboardBinding;
import com.jsb.calculator.databinding.DialogEachKeyBinding;

public class CustomizeKeyboardActivity extends AppCompatActivity {

    private ActivityCustomizeKeyboardBinding binding;
    private Activity activity;
    SharedPreferences sharedPreferences;

    private String keya = "a";
    private String keyb = "b";
    private String keyc = "c";
    private String keyd = "d";
    private String keye = "e";
    private String keyf = "f";
    private String keyg = "g";
    private String keyh = "h";
    private String keyi = "i";
    private String keyj = "j";
    private String keyk = "k";
    private String keyl = "l";
    private String keym = "m";
    private String keyn = "n";
    private String keyo = "o";
    private String keyp = "p";
    private String keyq = "q";
    private String keyr = "r";
    private String keys = "s";
    private String keyt = "t";
    private String keyu = "u";
    private String keyv = "v";
    private String keyw = "w";
    private String keyx = "x";
    private String keyy = "y";
    private String keyz = "z";

    private String keyA = "A";
    private String keyB = "B";
    private String keyC = "C";
    private String keyD = "D";
    private String keyE = "E";
    private String keyF = "F";
    private String keyG = "G";
    private String keyH = "H";
    private String keyI = "I";
    private String keyJ = "J";
    private String keyK = "K";
    private String keyL = "L";
    private String keyM = "M";
    private String keyN = "N";
    private String keyO = "O";
    private String keyP = "P";
    private String keyQ = "Q";
    private String keyR = "R";
    private String keyS = "S";
    private String keyT = "T";
    private String keyU = "U";
    private String keyV = "V";
    private String keyW = "W";
    private String keyX = "X";
    private String keyY = "Y";
    private String keyZ = "Z";

    private boolean isCapes = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityCustomizeKeyboardBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        activity = this;
        sharedPreferences = getSharedPreferences("com.jsb.calculator", MODE_PRIVATE);


        //def_abc


        binding.btCCaall.setEnabled(false);
        binding.btTextToSpeech.setEnabled(false);
        binding.btCopy.setEnabled(false);
        binding.btPaste.setEnabled(false);
        binding.btBackspace.setEnabled(false);
        binding.btQ123.setEnabled(false);
        binding.btComa.setEnabled(false);
        binding.btEmoji.setEnabled(false);
        binding.btSpace.setEnabled(false);
        binding.btDot2.setEnabled(false);
        binding.btKeyEnter.setEnabled(false);

        binding.btCaps.setOnClickListener(v -> {
            isCapes = !isCapes;
            setCaps(isCapes, isCapes);
        });


        //a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z=A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z

        setUpKeys(new Fonts().getFont(sharedPreferences.getString("def_abc", getString(R.string.font_1))));


        binding.btKey1.setOnClickListener(onKeyClickListener);
        binding.btKey2.setOnClickListener(onKeyClickListener);
        binding.btKey3.setOnClickListener(onKeyClickListener);
        binding.btKey4.setOnClickListener(onKeyClickListener);
        binding.btKey5.setOnClickListener(onKeyClickListener);
        binding.btKey6.setOnClickListener(onKeyClickListener);
        binding.btKey7.setOnClickListener(onKeyClickListener);
        binding.btKey8.setOnClickListener(onKeyClickListener);
        binding.btKey9.setOnClickListener(onKeyClickListener);
        binding.btKey10.setOnClickListener(onKeyClickListener);
        binding.btKey11.setOnClickListener(onKeyClickListener);
        binding.btKey12.setOnClickListener(onKeyClickListener);
        binding.btKey13.setOnClickListener(onKeyClickListener);
        binding.btKey14.setOnClickListener(onKeyClickListener);
        binding.btKey15.setOnClickListener(onKeyClickListener);
        binding.btKey16.setOnClickListener(onKeyClickListener);
        binding.btKey17.setOnClickListener(onKeyClickListener);
        binding.btKey18.setOnClickListener(onKeyClickListener);
        binding.btKey19.setOnClickListener(onKeyClickListener);
        binding.btKey20.setOnClickListener(onKeyClickListener);
        binding.btKey21.setOnClickListener(onKeyClickListener);
        binding.btKey22.setOnClickListener(onKeyClickListener);
        binding.btKey23.setOnClickListener(onKeyClickListener);
        binding.btKey24.setOnClickListener(onKeyClickListener);
        binding.btKey25.setOnClickListener(onKeyClickListener);
        binding.btKey26.setOnClickListener(onKeyClickListener);



    }

    View.OnClickListener onKeyClickListener = new OnClickListener() {
        @Override
        public void onClick(View v) {
            TextView textView = (TextView) v;
            showDialog((String) textView.getTag(R.string.small), (String) textView.getTag(R.string.big));
        }
    };

    private void showDialog(String oldSmall, String oldBig){
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setTitle("Change Key");
        DialogEachKeyBinding keyBinding = DialogEachKeyBinding.inflate(getLayoutInflater());

        builder.setView(keyBinding.getRoot());

        keyBinding.lket.setText(oldSmall);
        keyBinding.uket.setText(oldBig);

        builder.setPositiveButton("SAVE", (dialog, which) -> {

            String upper = keyBinding.uket.getText().toString();
            String lover = keyBinding.lket.getText().toString();
            if (!lover.isEmpty() && !upper.isEmpty() && !lover.contains(",") && !lover.contains("=") && !upper.contains(",") && !upper.contains("=")){

                String s = sharedPreferences.getString("def_abc", getString(R.string.font_1));
                s = s.replace(oldSmall, lover);
                s = s.replace(oldBig, upper);

                sharedPreferences.edit().putString("def_abc", s).apply();
                setUpKeys(new Fonts().getFont(sharedPreferences.getString("def_abc", getString(R.string.font_1))));

            }else {
                Toast.makeText(activity, "Invalid key", Toast.LENGTH_SHORT).show();
            }


            //m_Text = input.getText().toString();
        });
        builder.setNegativeButton(android.R.string.cancel, (dialog, which) -> dialog.cancel());

        builder.show();
    }

    private void setCaps(boolean caps, boolean a){
        setCaps(caps);
        if (a){
            binding.btCaps.setImageResource(R.drawable.ic_caps2);
        }else {
            binding.btCaps.setImageResource(R.drawable.ic_caps1);
        }
    }

    private void setCaps(boolean caps){
        if (caps){
            binding.btKey1.setText(keyQ);
            binding.btKey2.setText(keyW);
            binding.btKey3.setText(keyE);
            binding.btKey4.setText(keyR);
            binding.btKey5.setText(keyT);
            binding.btKey6.setText(keyY);
            binding.btKey7.setText(keyU);
            binding.btKey8.setText(keyI);
            binding.btKey9.setText(keyO);
            binding.btKey10.setText(keyP);
            binding.btKey11.setText(keyA);
            binding.btKey12.setText(keyS);
            binding.btKey13.setText(keyD);
            binding.btKey14.setText(keyF);
            binding.btKey15.setText(keyG);
            binding.btKey16.setText(keyH);
            binding.btKey17.setText(keyJ);
            binding.btKey18.setText(keyK);
            binding.btKey19.setText(keyL);
            binding.btKey20.setText(keyZ);
            binding.btKey21.setText(keyX);
            binding.btKey22.setText(keyC);
            binding.btKey23.setText(keyV);
            binding.btKey24.setText(keyB);
            binding.btKey25.setText(keyN);
            binding.btKey26.setText(keyM);
        }else {
            binding.btKey1.setText(keyq);
            binding.btKey2.setText(keyw);
            binding.btKey3.setText(keye);
            binding.btKey4.setText(keyr);
            binding.btKey5.setText(keyt);
            binding.btKey6.setText(keyy);
            binding.btKey7.setText(keyu);
            binding.btKey8.setText(keyi);
            binding.btKey9.setText(keyo);
            binding.btKey10.setText(keyp);
            binding.btKey11.setText(keya);
            binding.btKey12.setText(keys);
            binding.btKey13.setText(keyd);
            binding.btKey14.setText(keyf);
            binding.btKey15.setText(keyg);
            binding.btKey16.setText(keyh);
            binding.btKey17.setText(keyj);
            binding.btKey18.setText(keyk);
            binding.btKey19.setText(keyl);
            binding.btKey20.setText(keyz);
            binding.btKey21.setText(keyx);
            binding.btKey22.setText(keyc);
            binding.btKey23.setText(keyv);
            binding.btKey24.setText(keyb);
            binding.btKey25.setText(keyn);
            binding.btKey26.setText(keym);
        }



        binding.btKey1.setTag(R.string.small,keyq);
        binding.btKey2.setTag(R.string.small,keyw);
        binding.btKey3.setTag(R.string.small,keye);
        binding.btKey4.setTag(R.string.small,keyr);
        binding.btKey5.setTag(R.string.small,keyt);
        binding.btKey6.setTag(R.string.small,keyy);
        binding.btKey7.setTag(R.string.small,keyu);
        binding.btKey8.setTag(R.string.small,keyi);
        binding.btKey9.setTag(R.string.small,keyo);
        binding.btKey10.setTag(R.string.small,keyp);
        binding.btKey11.setTag(R.string.small,keya);
        binding.btKey12.setTag(R.string.small,keys);
        binding.btKey13.setTag(R.string.small,keyd);
        binding.btKey14.setTag(R.string.small,keyf);
        binding.btKey15.setTag(R.string.small,keyg);
        binding.btKey16.setTag(R.string.small,keyh);
        binding.btKey17.setTag(R.string.small,keyj);
        binding.btKey18.setTag(R.string.small,keyk);
        binding.btKey19.setTag(R.string.small,keyl);
        binding.btKey20.setTag(R.string.small,keyz);
        binding.btKey21.setTag(R.string.small,keyx);
        binding.btKey22.setTag(R.string.small,keyc);
        binding.btKey23.setTag(R.string.small,keyv);
        binding.btKey24.setTag(R.string.small,keyb);
        binding.btKey25.setTag(R.string.small,keyn);
        binding.btKey26.setTag(R.string.small,keym);

        binding.btKey1.setTag(R.string.big,keyQ);
        binding.btKey2.setTag(R.string.big,keyW);
        binding.btKey3.setTag(R.string.big,keyE);
        binding.btKey4.setTag(R.string.big,keyR);
        binding.btKey5.setTag(R.string.big,keyT);
        binding.btKey6.setTag(R.string.big,keyY);
        binding.btKey7.setTag(R.string.big,keyU);
        binding.btKey8.setTag(R.string.big,keyI);
        binding.btKey9.setTag(R.string.big,keyO);
        binding.btKey10.setTag(R.string.big,keyP);
        binding.btKey11.setTag(R.string.big,keyA);
        binding.btKey12.setTag(R.string.big,keyS);
        binding.btKey13.setTag(R.string.big,keyD);
        binding.btKey14.setTag(R.string.big,keyF);
        binding.btKey15.setTag(R.string.big,keyG);
        binding.btKey16.setTag(R.string.big,keyH);
        binding.btKey17.setTag(R.string.big,keyJ);
        binding.btKey18.setTag(R.string.big,keyK);
        binding.btKey19.setTag(R.string.big,keyL);
        binding.btKey20.setTag(R.string.big,keyZ);
        binding.btKey21.setTag(R.string.big,keyX);
        binding.btKey22.setTag(R.string.big,keyC);
        binding.btKey23.setTag(R.string.big,keyV);
        binding.btKey24.setTag(R.string.big,keyB);
        binding.btKey25.setTag(R.string.big,keyN);
        binding.btKey26.setTag(R.string.big,keyM);
    }

    private void setUpKeys(Fonts.Font font) {
        keya = font.a;
        keyb = font.b;
        keyc = font.c;
        keyd = font.d;
        keye = font.e;
        keyf = font.f;
        keyg = font.g;
        keyh = font.h;
        keyi = font.i;
        keyj = font.j;
        keyk = font.k;
        keyl = font.l;
        keym = font.m;
        keyn = font.n;
        keyo = font.o;
        keyp = font.p;
        keyq = font.q;
        keyr = font.r;
        keys = font.s;
        keyt = font.t;
        keyu = font.u;
        keyv = font.v;
        keyw = font.w;
        keyx = font.x;
        keyy = font.y;
        keyz = font.z;

        keyA = font.A;
        keyB = font.B;
        keyC = font.C;
        keyD = font.D;
        keyE = font.E;
        keyF = font.F;
        keyG = font.G;
        keyH = font.H;
        keyI = font.I;
        keyJ = font.J;
        keyK = font.K;
        keyL = font.L;
        keyM = font.M;
        keyN = font.N;
        keyO = font.O;
        keyP = font.P;
        keyQ = font.Q;
        keyR = font.R;
        keyS = font.S;
        keyT = font.T;
        keyU = font.U;
        keyV = font.V;
        keyW = font.W;
        keyX = font.X;
        keyY = font.Y;
        keyZ = font.Z;
        setCaps(isCapes, isCapes);
    }

}

File: app/src/main/java/com/jsb/calculator/activity/CustomizeKeyboardActivity.kt
package com.jsb.calculator.activity

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Bundle
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.jsb.calculator.R
import com.jsb.calculator.databinding.ActivityCustomizeKeyboardBinding
import com.jsb.calculator.databinding.DialogEachKeyBinding
import com.jsb.calculator.manager.LocalStorage
import com.jsb.calculator.modules.MyFont

class CustomizeKeyboardActivity : AppCompatActivity() {
    
    private lateinit var binding: ActivityCustomizeKeyboardBinding
    
    private var isCapes = false
    private lateinit var font: MyFont
    
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCustomizeKeyboardBinding.inflate(
            layoutInflater
        )
        setContentView(binding.root)
        binding.btCCaall.isEnabled = false
        binding.btTextToSpeech.isEnabled = false
        binding.btCopy.isEnabled = false
        binding.btPaste.isEnabled = false
        binding.btBackspace.isEnabled = false
        binding.btQ123.isEnabled = false
        binding.btComa.isEnabled = false
        binding.btEmoji.isEnabled = false
        binding.btSpace.isEnabled = false
        binding.btDot2.isEnabled = false
        binding.btKeyEnter.isEnabled = false
        binding.btCaps.setOnClickListener {
            isCapes = !isCapes
            setCaps(isCapes)
        }
        font = LocalStorage(this).getFont()
        setCaps(isCapes)
        addOnClick(binding.btKey1, "q")
        addOnClick(binding.btKey2, "w")
        addOnClick(binding.btKey3, "e")
        addOnClick(binding.btKey4, "r")
        addOnClick(binding.btKey5, "t")
        addOnClick(binding.btKey6, "y")
        addOnClick(binding.btKey7, "u")
        addOnClick(binding.btKey8, "i")
        addOnClick(binding.btKey9, "o")
        addOnClick(binding.btKey10, "p")
        addOnClick(binding.btKey11, "a")
        addOnClick(binding.btKey12, "s")
        addOnClick(binding.btKey13, "d")
        addOnClick(binding.btKey14, "f")
        addOnClick(binding.btKey15, "g")
        addOnClick(binding.btKey16, "h")
        addOnClick(binding.btKey17, "j")
        addOnClick(binding.btKey18, "k")
        addOnClick(binding.btKey19, "l")
        addOnClick(binding.btKey20, "z")
        addOnClick(binding.btKey21, "x")
        addOnClick(binding.btKey22, "c")
        addOnClick(binding.btKey23, "v")
        addOnClick(binding.btKey24, "b")
        addOnClick(binding.btKey25, "n")
        addOnClick(binding.btKey26, "a")
    }

    private fun addOnClick(bt: TextView, name: String) {
        bt.setOnClickListener {
            showDialog(name)
        }
    }


    private fun showDialog(name: String) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Change Key")
        val keyBinding = DialogEachKeyBinding.inflate(
            layoutInflater
        )

        val field = font.javaClass.getDeclaredField(name)
        field.isAccessible = true
        val bigField = font.javaClass.getDeclaredField("big${name.uppercase()}")
        bigField.isAccessible = true
        keyBinding.lket.setText(field.get(font) as String)
        keyBinding.uket.setText(bigField.get(font) as String)
        builder.setView(keyBinding.root)
        builder.setPositiveButton("SAVE") { dialog: DialogInterface?, which: Int ->
            val upper = keyBinding.uket.text.toString()
            val lover = keyBinding.lket.text.toString()
            if (lover.isNotEmpty() && upper.isNotEmpty() && !lover.contains(",") && !lover.contains("=") && !upper.contains(
                    ","
                ) && !upper.contains("=")
            ) {

                field.set(font, lover)
                bigField.set(font, upper)

                LocalStorage(this).saveFont(font)
                setCaps(isCapes)
            } else {
                Toast.makeText(this, "Invalid key", Toast.LENGTH_SHORT).show()
            }
        }
        builder.setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, which: Int -> dialog.cancel() }
        builder.show()
    }

    private fun setCaps(caps: Boolean) {
        if (caps) {
            binding.btCaps.setImageResource(R.drawable.ic_caps2)
            binding.btKey1.text = font.bigQ
            binding.btKey2.text = font.bigW
            binding.btKey3.text = font.bigE
            binding.btKey4.text = font.bigR
            binding.btKey5.text = font.bigT
            binding.btKey6.text = font.bigY
            binding.btKey7.text = font.bigU
            binding.btKey8.text = font.bigI
            binding.btKey9.text = font.bigO
            binding.btKey10.text = font.bigP
            binding.btKey11.text = font.bigA
            binding.btKey12.text = font.bigS
            binding.btKey13.text = font.bigD
            binding.btKey14.text = font.bigF
            binding.btKey15.text = font.bigG
            binding.btKey16.text = font.bigH
            binding.btKey17.text = font.bigJ
            binding.btKey18.text = font.bigK
            binding.btKey19.text = font.bigL
            binding.btKey20.text = font.bigZ
            binding.btKey21.text = font.bigX
            binding.btKey22.text = font.bigC
            binding.btKey23.text = font.bigV
            binding.btKey24.text = font.bigB
            binding.btKey25.text = font.bigN
            binding.btKey26.text = font.bigM
        } else {
            binding.btCaps.setImageResource(R.drawable.ic_caps1)
            binding.btKey1.text = font.q
            binding.btKey2.text = font.w
            binding.btKey3.text = font.e
            binding.btKey4.text = font.r
            binding.btKey5.text = font.t
            binding.btKey6.text = font.y
            binding.btKey7.text = font.u
            binding.btKey8.text = font.i
            binding.btKey9.text = font.o
            binding.btKey10.text = font.p
            binding.btKey11.text = font.a
            binding.btKey12.text = font.s
            binding.btKey13.text = font.d
            binding.btKey14.text = font.f
            binding.btKey15.text = font.g
            binding.btKey16.text = font.h
            binding.btKey17.text = font.j
            binding.btKey18.text = font.k
            binding.btKey19.text = font.l
            binding.btKey20.text = font.z
            binding.btKey21.text = font.x
            binding.btKey22.text = font.c
            binding.btKey23.text = font.v
            binding.btKey24.text = font.b
            binding.btKey25.text = font.n
            binding.btKey26.text = font.m
        }
    }
}